### PR TITLE
DevTools Deprecation Module Improvements (ED-620)

### PR DESCRIPTION
**Pull request by udidol**
DevTools Deprecation Module: 
- Added a `do_deprecated_action()` method.
- Added a `run_compatible_method()` method which detects whether a method belongs to the current instance or to its parent and runs the compatible method.

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### CODE REVIEW

File: modules/dev-tools/deprecation.php [hunk 1]
```diff
@@ -259,4 +259,23 @@ public function deprecated_argument( $function, $version ) {
...
+	 * @param string $hook
+	 * @param array $args
+	 * @param string $version
+	 * @param string $replacement
+	 * @param null|string $base_version
+	 *
+	 * @throws \Exception
+	 */
+	public function do_deprecated_action( $hook, $args, $version, $replacement = '', $base_version = null ) {
+		if ( ! has_action( $hook  ) ) {
```
Comment (rodolphebertozzo) [inline modules/dev-tools/deprecation.php:273]: if ( ! has_action( $hook ) ) {

File: modules/dev-tools/deprecation.php [hunk 2]
```diff
@@ -259,4 +259,23 @@ public function deprecated_argument( $function, $version ) {
...
+	/**
+	 * @param string $hook
+	 * @param array $args
+	 * @param string $version
+	 * @param string $replacement
+	 * @param null|string $base_version
+	 *
+	 * @throws \Exception
+	 */
+	public function do_deprecated_action( $hook, $args, $version, $replacement = '', $base_version = null ) {
```
Comment (kobizz) [inline modules/dev-tools/deprecation.php:272]: Add a description and `@since`

**Review by kobizz — CHANGES_REQUESTED** [review]

File: modules/dev-tools/deprecation.php (hunk 2 above)
Comment (udidol) [reply modules/dev-tools/deprecation.php:296]: Done.

**Review by kobizz — DISMISSED** [review]

File: tests/phpunit/elementor/modules/dev-tools/test-deprecation.php [hunk 3]
```diff
@@ -205,4 +205,36 @@ public function test_deprecated_function_hard() {
...
+		ob_start();
+
+		$this->deprecation->do_deprecated_action( 'elementor/test/deprecated_action', [], '0.0.0', '', '0.5.0' );
+
+		$result = ob_get_clean();
+
+		$this->assertEquals( $result, 'Testing Do Deprecated Action' );
+	}
+
+	public function test_do_deprecated_action_soft() {
```
Comment (matipojo) [inline tests/phpunit/elementor/modules/dev-tools/test-deprecation.php:225]:
```suggestion
	public function test_do_deprecated__action_soft() {
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/modules/dev-tools/test-deprecation.php (hunk 3 above)
Comment (udidol) [reply tests/phpunit/elementor/modules/dev-tools/test-deprecation.php:225]: why?

File: modules/dev-tools/assets/js/editor/dev-tools.js [hunk 4]
```diff
@@ -7,10 +7,8 @@ export default class extends elementorModules.editor.utils.Module {
 		// eslint-disable-next-line camelcase
 		const notices = elementor.config.dev_tools?.deprecation.soft_notices;
```
Comment (kobizz) [inline modules/dev-tools/assets/js/editor/dev-tools.js:8]:
```suggestion
		const notices = elementor.config.dev_tools.deprecation.soft_notices;
```

**Review by kobizz — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/modules/dev-tools/test-deprecation.php (hunk 3 above)
Comment (matipojo) [reply tests/phpunit/elementor/modules/dev-tools/test-deprecation.php:224]: I fixed it, it the method name should be separated from the use case.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/modules/dev-tools/test-deprecation.php (hunk 3 above)
Comment (udidol) [reply tests/phpunit/elementor/modules/dev-tools/test-deprecation.php:224]: I Don't understand @matipojo

File: modules/dev-tools/deprecation.php [hunk 5]
```diff
@@ -188,11 +200,13 @@ private function check_deprecation( $entity, $version, $replacement, $base_versi
...
 		if ( defined( 'WP_DEBUG' ) && WP_DEBUG && $diff <= self::SOFT_VERSIONS_COUNT ) {
 			// Soft deprecated.
-			$this->soft_deprecated_notices [] = [
-				$entity,
-				$version,
-				$replacement,
-			];
+			if ( ! isset( $this->soft_deprecated_notices[ $entity ] ) ) {
+				$this->soft_deprecated_notices[ $entity ] = [
+					$entity,
```
[the remark this answers is not in the document]
Comment (udidol) [reply modules/dev-tools/deprecation.php]: Done.

File: modules/dev-tools/assets/js/editor/dev-tools.js (hunk 4 above)
Comment (udidol) [reply modules/dev-tools/assets/js/editor/dev-tools.js]: Done.

Comment (udidol) [reply modules/dev-tools/assets/js/editor/dev-tools.js]: Done.

**Review by kobizz — CHANGES_REQUESTED** [review]
Fix JS-Lint issue

File: tests/phpunit/elementor/modules/dev-tools/test-deprecation.php (hunk 3 above)
Comment (udidol) [reply tests/phpunit/elementor/modules/dev-tools/test-deprecation.php:224]: Changed it to double underscore.

**Review by kobizz — DISMISSED** [review]

**Review by matipojo — APPROVED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
Please fix the error/test on;
https://github.com/elementor/elementor/pull/12923/checks?check_run_id=1352112030

**Review by kobizz — DISMISSED** [review]

File: tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js [hunk 6]
```diff
@@ -2,12 +2,12 @@ QUnit.module( 'File: modules/dev-tools/assets/js/editor/dev-tools.js', () => {
...
 
-		elementor.config.dev_tools = {
-			deprecation: { soft_notices: [ [ 'name', 'version', 'replacement' ] ] },
-		};
+		// The soft notices object structure changed, instead of a nested array, it is now an object, for caching.
+		elementor.config.dev_tools.deprecation.soft_notices = { test: [ 'version', 'replacement' ] };
 
 		elementorCommon.helpers.softDeprecated = ( ... args ) => {
-			assert.deepEqual( [ args ], elementor.config.dev_tools.deprecation.soft_notices );
+			// args[ 0 ] is the deprecated function name, args[ 1 ] is the version number, args[ 2 ] is the replacement method.
```
Comment (matipojo) [inline tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js:9]: Remove comment.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js (hunk 6 above)
Comment (udidol) [reply tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js:9]: Done.

**Review by rodolphebertozzo — APPROVED** [review]

**Review by matipojo — APPROVED** [review]

**Review by kobizz — APPROVED** [review]

**Pull request merged**